File: backend/app/schemas/preferences.py
import json
from typing import Literal

from pydantic import BaseModel, Field, field_validator


PreferenceFactor = Literal["accessibility", "walk", "transfers"]
DEFAULT_PRIORITY_ORDER: list[PreferenceFactor] = ["accessibility", "walk", "transfers"]


class TravelPreferences(BaseModel):
    # Defaults keep unconfigured planning aligned with Google Maps fastest transit.
    accessibility_first: bool = False
    least_walk: bool = False
    fewest_transfers: bool = False
    priority_order: list[PreferenceFactor] = Field(default_factory=lambda: list(DEFAULT_PRIORITY_ORDER))

    @field_validator("priority_order", mode="before")
    @classmethod
    def normalize_priority_order(cls, value) -> list[PreferenceFactor]:
        if isinstance(value, str):
            try:
                value = json.loads(value)
            except json.JSONDecodeError:
                value = None
        if not isinstance(value, list):
            return list(DEFAULT_PRIORITY_ORDER)

        normalized: list[PreferenceFactor] = []
        for item in value:
            if item in DEFAULT_PRIORITY_ORDER and item not in normalized:
                normalized.append(item)

        for factor in DEFAULT_PRIORITY_ORDER:
            if factor not in normalized:
                normalized.append(factor)

        return normalized

File: frontend/src/types/preferences.ts
export type PreferenceFactor = 'accessibility' | 'walk' | 'transfers';

export const DEFAULT_PRIORITY_ORDER: PreferenceFactor[] = ['accessibility', 'walk', 'transfers'];

export interface TravelPreferences {
  accessibilityFirst: boolean;
  leastWalk: boolean;
  fewestTransfers: boolean;
  priorityOrder: PreferenceFactor[];
}

export interface ApiTravelPreferences {
  accessibility_first: boolean;
  least_walk: boolean;
  fewest_transfers: boolean;
  priority_order?: PreferenceFactor[];
}

export function normalizePriorityOrder(value?: PreferenceFactor[] | null): PreferenceFactor[] {
  const normalized: PreferenceFactor[] = [];
  if (Array.isArray(value)) {
    for (const item of value) {
      if (DEFAULT_PRIORITY_ORDER.includes(item) && !normalized.includes(item)) {
        normalized.push(item);
      }
    }
  }
  for (const item of DEFAULT_PRIORITY_ORDER) {
    if (!normalized.includes(item)) normalized.push(item);
  }
  return normalized;
}

export function toApiPreferences(preferences: TravelPreferences): ApiTravelPreferences {
  // Boundary mapper between frontend camelCase and backend snake_case.
  return {
    accessibility_first: preferences.accessibilityFirst,
    least_walk: preferences.leastWalk,
    fewest_transfers: preferences.fewestTransfers,
    priority_order: normalizePriorityOrder(preferences.priorityOrder)
  };
}

export function fromApiPreferences(preferences: ApiTravelPreferences): TravelPreferences {
  return {
    accessibilityFirst: preferences.accessibility_first,
    leastWalk: preferences.least_walk,
    fewestTransfers: preferences.fewest_transfers,
    priorityOrder: normalizePriorityOrder(preferences.priority_order)
  };
}
